File: codility/currency-conversion.js
class Graph {
    constructor() {
        this.fx_rates = {};
        this.max_exchange_rate = -1.0;
    }

    add_edge(from_currency, to_currency, ex_rate) {
        if(!(from_currency in this.fx_rates)) {
            this.fx_rates[from_currency] = [];
        } 
        this.fx_rates[from_currency].push([to_currency, ex_rate]);
    }

    dfs(current_currency, target_currency, visited, path, amount) {
        visited[current_currency] = true;
        path.push(amount)
    
        if(current_currency === target_currency) {
            let current_amount = path.reduce((a, b) => a * b, 1);

            if(current_amount !== 1.0 && current_amount > this.max_exchange_rate) {
                this.max_exchange_rate = current_amount;
            }

            // console.log(path, current_amount);
        }
        else {
            for(const [currency, rate] of this.fx_rates[current_currency]) {
                if(!visited[currency]) {
                    this.dfs(currency, target_currency, visited, path, rate);
                }
            }
        }

        path.pop();
        visited[current_currency] = false;
    }

    traverse_path(original_currency, target_currency) {
        let visited = {}, path = [];

        for(const currency of Object.keys(this.fx_rates)) {
            visited[currency] = false;
        }

        this.dfs(original_currency, target_currency, visited, path, 1.0);

        // const m = Number((Math.abs(this.max_exchange_rate) * 100).toPrecision(15));
        // return Math.round(m) / 100 * Math.sign(this.max_exchange_rate);
        return this.max_exchange_rate.toFixed(2);
    }
}

const generate_graph = fx_rates => {
    let graph_obj = new Graph();
    hash_table = {};

    for(const [from_currency, to_currency, rate] of fx_rates) {
        graph_obj.add_edge(from_currency, to_currency, rate);
        hash_table[`${from_currency}-${to_currency}`] = rate;
    }
       
    for(const [from_currency, to_currency, rate] of fx_rates) {
        if(!(`${to_currency}-${from_currency}` in hash_table)) {
            graph_obj.add_edge(to_currency, from_currency, 1 / rate);
        }
    }
       
    return graph_obj;
};


(function () {
    const inputs = [
        {
            "fx_rates": [
                ['USD', 'JPY', 110],
                ['USD', 'AUD', 1.45],
                ['JPY', 'GBP', 0.0070]

            ],
            "original_currency": 'GBP',
            "target_currency": 'AUD',
            "expected": 1.88
        },
        {
            "fx_rates": [
                ['USD', 'JPY', 109],
                ['USD', 'GBP', 0.71],
                ['GBP', 'JPY', 155]
            ],
            "original_currency": 'USD',
            "target_currency": 'JPY',
            "expected": 110.05
        },
        {
            "fx_rates": [
                ['USD', 'GBP', 0.7],
                ['USD', 'JPY', 109],
                ['GBP', 'JPY', 155],
                ['CAD', 'CNY', 5.27],
                ['CAD', 'KRW', 921]

            ],
            "original_currency": 'USD',
            "target_currency": 'CNY',
            "expected": -1.0
        },
        {
            "fx_rates": [
                ['USD', 'CAD', 1.3],
                ['USD', 'GBP', 0.71],
                ['USD', 'JPY', 109],
                ['GBP', 'JPY', 155]

            ],
            "original_currency": 'USD',
            "target_currency": 'JPY',
            "expected": 110.05
        },
        {
            "fx_rates": [
                ['MVR', 'BDT', 5.17728],
                ['GYD', 'BDT', 0.39521],
                ['BDT', 'MVR', 0.17772],
                ['BDT', 'GYD', 2.43054]

            ],
            "original_currency": 'MVR',
            "target_currency": 'GYD',
            "expected": 12.58
        },
        {
            "fx_rates": [
                ['USD', 'EUR', 0.9],
                ['EUR', 'CAD', 1.46]

            ],
            "original_currency": 'CAD',
            "target_currency": 'USD',
            "expected": 0.76
        },
        {
            "fx_rates": [
                ['USD', 'CAD', 0.5],
                ['GBP', 'USD', 0.5],
                ['CAD', 'JPY', 1.75],
                ['JPY', 'GBP', 3],
                ['EUR', 'JPY', 1 / 5],
                ['BDT', 'JPY', 4],
                ['JPY', 'BDT', 0.3],
                ['GBP', 'EUR', 10],
                ['GBP', 'BDT', 3]
            ],
            "original_currency": 'USD',
            "target_currency": 'EUR',
            "expected": 120.0
        }
    ];

    let test_passed = 0, i = 0;
    for(const val of inputs) {
        const g = generate_graph(val["fx_rates"]);
        const output = g.traverse_path(val["original_currency"], val["target_currency"]);
        if(output == val['expected']) {
            console.log(`${i}. CORRECT Answer\nOutput:${output}\nExpected:${val['expected']}\n\n`);
            test_passed += 1;
        }
        else {
            console.log(`${i}. WRONG Answer\nOutput:${output}\nExpected:${val['expected']}\n\n`)
        }
        i += 1;
    }
       
    console.log(`Passed - ${test_passed}/${i}\n`);
})();

File: neo-financial/Currency-Conversion.py
import json
import csv
from urllib.request import urlopen


class Graph:
    """
    @Title: Neo Financial - Technical Assignment
    @Problem: Currency Conversion
    @Author: Tahmid Tanzim <tahmid.tanzim@gmail.com>
    @GitHub: https://github.com/tahmid-tanzim
    @Date: 2022-04-27
    """

    def __init__(self, items):
        self.fx_rates = dict()              # Adjacency List {'EUR': [('USD', 12.3), ('CAD', 45.6)]}
        self.currency_list = list()         # [{'code': 'EUR', 'name': 'Euro'}]
        self.visited = dict()               # {'EUR': False}
        self.max_exchange_rate = -1.0
        self.best_conversion_path = list()  # ['CAD', 'BRL', 'ARS']
        self.setup(items)

    def addEdge(self, from_currency_code, to_currency_code, from_currency_name, to_currency_name, ex_rate):
        if from_currency_code not in self.fx_rates:
            self.fx_rates[from_currency_code] = list()
            self.currency_list.append({
                "code": from_currency_code,
                "name": from_currency_name
            })
        if to_currency_code not in self.fx_rates:
            self.fx_rates[to_currency_code] = list()
            self.currency_list.append({
                "code": to_currency_code,
                "name": to_currency_name
            })

        # Creating Adjacency List
        self.fx_rates[from_currency_code].append((to_currency_code, ex_rate,))

    def setup(self, items):
        hash_table = dict()
        for item in items:
            self.addEdge(
                item['fromCurrencyCode'],
                item['toCurrencyCode'],
                item['fromCurrencyName'],
                item['toCurrencyName'],
                item['exchangeRate']
            )
            hash_table[f"{item['fromCurrencyCode']}-{item['toCurrencyCode']}"] = item['exchangeRate']

        # Store reversed exchange rate if not provided as an input
        for item in items:
            if f"{item['toCurrencyCode']}-{item['fromCurrencyCode']}" not in hash_table:
                self.addEdge(
                    item['toCurrencyCode'],
                    item['fromCurrencyCode'],
                    item['toCurrencyName'],
                    item['fromCurrencyName'],
                    float(1 / item['exchangeRate'])
                )

    def depthFirstSearch(self, current_currency, target_currency, amount, conversion_path, currency_stack):
        # Mark current currency as visited
        self.visited[current_currency] = True

        # Pre-calculated conversion rate stored in currency_stack
        # Top of the stack is the last calculated conversion rate
        currency_stack.append(currency_stack[-1] * amount)
        conversion_path.append(current_currency)

        # Check if current_currency reached to the destination.
        if current_currency == target_currency:
            # print(currency_stack, conversion_path)
            current_amount = currency_stack[-1]
            if current_amount > self.max_exchange_rate:
                self.max_exchange_rate = current_amount
                self.best_conversion_path = conversion_path.copy()
        else:
            # If current_currency NOT reached to the destination.
            # Then recursively check for all the adjacent NOT visited currency
            for adjacent_currency, rate in self.fx_rates[current_currency]:
                if not self.visited[adjacent_currency]:
                    self.depthFirstSearch(adjacent_currency, target_currency, rate, conversion_path, currency_stack)

        # Reset before return from DFS
        currency_stack.pop()
        conversion_path.pop()
        self.visited[current_currency] = False

    def startPathTraversal(self, original_currency, target_currency, amount):
        # initialize max_exchange_rate & best_conversion_path
        self.max_exchange_rate = -1.0
        self.best_conversion_path = list()

        # initialize visited dictionary
        for currency in self.currency_list:
            code = currency["code"]
            self.visited[code] = False

        self.depthFirstSearch(original_currency, target_currency, amount, conversion_path=[], currency_stack=[1.0])
        return self.max_exchange_rate, " | ".join(self.best_conversion_path),


def generateCSV(row, version):
    # CSV file name with seedId as version number.
    filename = f"Currency-Conversion-Output-{version}.csv"

    # Append a row to the CSV file
    with open(filename, 'a', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(row)
        file.close()


def getResponse(url):
    response = None
    try:
        response = urlopen(url)
    except Exception as ex:
        print("Sorry!", ex)
        return list()
    else:
        body = response.read()
        return json.loads(body)
    finally:
        if response is not None:
            response.close()


if __name__ == '__main__':
    seedId = 62861
    data = getResponse(f"https://api-coding-challenge.neofinancial.com/currency-conversion?seed={seedId}")
    graph_obj = Graph(data)

    start_currency_code = "CAD"
    start_amount = 100.0

    for currency in graph_obj.currency_list:
        end_currency_code = currency["code"]
        end_currency_name = currency["name"]

        # Ignore operation, if end_currency_code is CAD
        if start_currency_code != end_currency_code:
            best_conversion_rate, pipe_delimited_path = graph_obj.startPathTraversal(
                start_currency_code,
                end_currency_code,
                start_amount
            )

            # Check if any path exist from start_currency_code to end_currency_code
            if best_conversion_rate != -1.0:
                # print(
                #     end_currency_code,
                #     end_currency_name,
                #     best_conversion_rate,
                #     pipe_delimited_path,
                #     end="\n\n"
                # )
                generateCSV([
                    end_currency_code,
                    end_currency_name,
                    best_conversion_rate,
                    pipe_delimited_path
                ], seedId)
